crates/amaru/src/stages/pull_p2p.rs

```diff
@@ -0,0 +1,348 @@
+        Point::Origin => pallas_network2::protocol::Point::Origin,
+        Point::Specific(slot, hash) => {
+            pallas_network2::protocol::Point::Specific(*slot, hash.clone())
+        }
+    }
+}
+
+fn pallas_to_amaru_point(point: &pallas_network2::protocol::Point) -> Point {
+    match point {
+        pallas_network2::protocol::Point::Origin => Point::Origin,
+        pallas_network2::protocol::Point::Specific(slot, hash) => {
+            Point::Specific(*slot, hash.clone())
+        }
+    }
+}
+
+fn pallas_to_amaru_peer(peer: &PeerId) -> amaru_kernel::peer::Peer {
+    amaru_kernel::peer::Peer::new(&peer.to_string())
+}
+
+#[derive(Default)]
+pub struct WorkUnit {
+    /// if set, instruct the network machinery to execute a particular command.
+    pub network_cmd: Option<InitiatorCommand>,
+
+    /// if set, dispatch a particular chain sync event to downstream stages.
+    pub sync_data: Option<ChainSyncEvent>,
+
+    /// if set, dispatch a particular block body data to downstream stages.
+    pub block_data: Option<BlockBodyData>,
+}
+
+impl From<InitiatorCommand> for WorkUnit {
+    fn from(command: InitiatorCommand) -> Self {
+        Self {
+            network_cmd: Some(command),
+            sync_data: None,
+            block_data: None,
+        }
+    }
+}
+
+/// A request from a downstream stage
+#[derive(Debug, Clone)]
+pub enum DownstreamRequest {
+    /// A request to fetch a particular block range.
+    FetchBlock(BlockRange),
+}
+
+pub struct Config {
+    /// The chain-sync intersection to ask to every hot peer.
+    pub intersection: Vec<Point>,
+
+    /// To configure the "promotion" behavior (eg: hot, warm, cold peers).
+    pub promotion: PromotionConfig,
+
+    /// The initial set of peers to connect to (aka: bootstrap nodes).
+    pub initial_peers: Vec<PeerId>,
+
+    /// The interval at which the housekeeping task is run (defaults to 3 secs).
+    pub housekeeping: Option<Duration>,
+}
+
+#[derive(Stage)]
+// TODO: revisit stage name. I would prefer something like "network", "p2p", etc. I'll leave the
+// decision for maintainers since it might affect tracing.
+#[stage(name = "stage.chain_sync_client", unit = "WorkUnit", worker = "Worker")]
+pub struct Stage {
+    /// To receive downstream requests.
+    pub requests: RequestPort,
+
+    /// To output chain sync data.
+    pub sync_data: SyncDataPort,
+
+    /// To output block body data.
+    pub block_data: BlockDataPort,
+
+    config: Config,
+}
+
+impl Stage {
+    pub fn new(config: Config) -> Self {
+        Self {
+            requests: Default::default(),
+            sync_data: Default::default(),
+            block_data: Default::default(),
+            config,
+        }
+    }
+}
+
+pub struct Worker {
+    pub network: Manager<TcpInterface, InitiatorBehavior, NetworkMessage>,
+    pub housekeeping: tokio::time::Interval,
+}
+
+impl Worker {
+    fn handle_network(&self, event: Option<InitiatorEvent>) -> WorkSchedule<WorkUnit> {
```

Comment:
passing an `Option<>` here feels a bit weird, why not avoid calling the function altogether if we don't have an event to handle?

Reply:
Yeah, I kind of agree. I did this to improve readability on the `select!` branches and emphasize the fact that returning `idle` for an empty event is a business logic decision (as opposed to glue code between functions).

One could make use of empty network ticks to provide metrics about the absence of network activity (eg: a counter of idle event).

Having said that, feel free to change it if the alternative feels more suitable.